### ISSUE TITLE
Add analyzer: Call async methods when in an async method

### ISSUE THREAD

**Issue by jeffhandley** [dotnet/runtime#45661]
[AB#1245428](https://devdiv.visualstudio.com/10e66e43-9645-4201-b128-0fdc3769cc17/_workitems/edit/1245428)
## Background and Motivation

The [VSTHRD103](https://github.com/microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD103.md) analyzer exists and it is broadly valuable. We should promote this analyzer to be part of the standard set of analyzers that ship with the .NET SDK.

## Examples of patterns that are flagged by this analyzer

```csharp
Task DoAsync()
{
    file.Read(buffer, 0, 10);
}
```

All methods where an Async-suffixed equivalent exists will produce this warning
when called from a `Task`-returning method.
In addition, calling `Task.Wait()`, `Task<T>.Result` or `Task.GetAwaiter().GetResult()`
will produce this warning.

## Solution

Await the async version of the method:

```csharp
async Task DoAsync()
{
    await file.ReadAsync(buffer, 0, 10);
}
```

**Comment by stephentoub** [comment]
This is worth pursuing in some fashion,  but it also has the potential to be very noisy,  e.g. it'll wreak havoc on the pattern of sharing one code path between async and sync code, e.g.
https://github.com/dotnet/runtime/blob/a5d0bcf19f778ac3cc579320c64adc4cf2e5f1d7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1304-1314

**Comment by jeffhandley** [comment]
I expect one of the first steps we'll take on this proposal is to run the existing analyzer against the runtime repo. From that, we can assess what the false positive rate would be and guide the path forward.

**Comment by carlossanlop** [comment]
> _This is worth pursuing in some fashion, but it also has the potential to be very noisy, e.g. it'll wreak havoc on the pattern of sharing one code path between async and sync code_

I think this analyzer offers great value. We can always set the category to Info or even Hidden to reduce the noise.

**Comment by carlossanlop** [comment]
Considering the [VSTHRD103](https://github.com/microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD103.md) analyzer already exists, we want to:

- Bring it to the API Proposal meeting to decide if we want to turn it on.
- In Parallel, test it in the dotnet/runtime repo to see how well it performs there. If we find any bugs or unhandled cases, address them.

Now, since this analyzer does not live in the dotnet/roslyn-analyzers repo, I'm unsure what the process would be to enable it, if we decide to do so.

@mavasani @AArnott can you help me answer the following questions?:

- I see [a large number of unit tests](https://github.com/microsoft/vs-threading/blob/master/test/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD103UseAsyncOptionAnalyzerTests.cs) for this analyzer. @AArnott do you remember if both the analyzer and fixer were [coded to completion](https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading.Analyzers.CodeFixes/VSTHRD103UseAsyncOptionCodeFix.cs), or only partially?
- If we approve it, @mavasani would we need to move its source code to dotnet/roslyn-analyzers, or would it stay in microsoft/vs-threading?

**Comment by AArnott** [comment]
From past discussions on this question, the preferred path is for dotnet to take a copy of the analyzer source code, change its ID to match your scheme, and continue to develop it. When dotnet's copy is publicly available via nuget package, we would consider removing the original one from vs-threading.

As far as its "completion" state, we were quite happy with it. There are [a few open issues](https://github.com/microsoft/vs-threading/issues?q=is%3Aissue+is%3Aopen+vsthrd103) that you should consider resolving in your version though.

**Comment by carlossanlop** [comment]
Thank you @AArnott. I think this is ready for review since the analyzer already exists and we can just need to move it from the other repo.

In parallel, we can run the existing one in the runtime repo, see how noisy it is.

- Suggested severity: Warning (if we discover it's not too noisy)
- Suggested milestone: 6.0
- Suggested category: Reliability
- Analyzer: Small (already exists, may need tweaking)
- Fixer: Small (already exists, may need tweaking)

**Comment by terrajobst** [comment]
* This might be noisy, at least in the runtime repo
* We should run this first so we can see what the noise is to make a determination of on/off by default and severity

```C#
namespace System.Linq
{
    public static partial class Enumerable
    {
        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index);
        public static TSource ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index);
        public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> source, Range range);
    }
    public static partial class Queryable
    {
        public static TSource ElementAt<TSource>(this IQueryable<TSource> source, Index index);
        public static TSource ElementAtOrDefault<TSource>(this IQueryable<TSource> source, Index index);
        public static IQueryable<TSource> Take<TSource>(this IQueryable<TSource> source, Range range);
    }
}
```

**Comment by carlossanlop** [comment]
I turned it on locally for the runtime repo and I got 259 Errors. I created a draft PR so we can have an official CI output linked to this issue: https://github.com/dotnet/runtime/pull/48540

@terrajobst @stephentoub @buyaa-n I think this analyzer should be enabled as `Info` for dotnet/runtime repo. It would also make sense to mark it as `Info` globally.

Should I bring this to API review again, or would it be ok to have it approved offline?

**Comment by carlossanlop** [comment]
@terrajobst based on [this comment](https://github.com/dotnet/runtime/pull/48540/files#r580719802) by Stephen, can you please mark this API as approved? The analyzer will be disabled by default in the runtime repo, and the default severity will be info.

**Comment by jeffhandley** [comment]
We are targeting 6.0 RC1 for this analyzer & fixer.

**Comment by jeffhandley** [comment]
Closed via dotnet/roslyn-analyzers#5367
